### PR TITLE
fix: Prompt for credentials when 401 in editor, fix profile propagation

### LINKED ISSUE #3197: V3 doesn't prompt for credentials when opening file in editor
https://github.com/user-attachments/assets/68752325-5d53-4beb-9a60-b91c21c6f867

_Originally posted by @JillieBeanSim in https://github.com/zowe/zowe-explorer-vscode/pull/3157#pullrequestreview-2354573221_

### PR DESCRIPTION
**Pull request by traeok**
## Proposed changes

Fixes #3197, #3289 

### How to test

- Checkout this branch in the repo with VS Code opened
- `pnpm install && pnpm build`
- After the build has finished, click on the Run & Debug panel on the VS Code Side Bar
- Select "Run VS Code Extension" from the tasks list at the top and select the Play ▶ button
- Once opened, make sure that you have (or set) valid credentials on a profile in ZE
- Execute a search on that profile and expand a folder so that a file/PDS member/spool item is visible
- Before opening the file, right click on your profile -> "Manage Profile" -> "Update Credentials"
- Enter in invalid credentials. I recommend using a fake username to prevent lockouts
- Once finished, try to open the file. Notice the error in the editor appears but you are prompted to update your credentials.
- Enter in the valid credentials
- Once done, click "Try again" in the editor to fetch & open the file

## Release Notes

Milestone: 3.0.3?

Changelog:

- Fixed issue where users were not prompted to enter credentials if a 401 error was encountered when opening files, data sets or spools in the editor. [#3197](https://github.com/zowe/zowe-explorer-vscode/issues/3197)
- Fixed issue where profile credential updates or token changes were not reflected within the filesystem. [#3289](https://github.com/zowe/zowe-explorer-vscode/issues/3289)

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (non-breaking change which adds or improves functionality)
- [ ] Breaking change (a change that would cause existing functionality to not work as expected)
- [ ] Documentation (Markdown, README updates)
- [ ] Other (please specify above in "Proposed changes" section)

## Checklist

**General**

- [x] I have read the [CONTRIBUTOR GUIDANCE](https://github.com/zowe/zowe-explorer-vscode/wiki/Best-Practices:-Contributor-Guidance) wiki
- [x] All PR dependencies have been merged and published (if applicable)
- [ ] A GIF or screenshot is included in the PR for visual changes
- [x] The pre-publish command has been executed:
  - **v2 and below:** `yarn workspace vscode-extension-for-zowe vscode:prepublish`
  - **v3:** `pnpm --filter vscode-extension-for-zowe vscode:prepublish`

**Code coverage**

- [x] There is coverage for the code that I have added
- [x] I have added new test cases and they are passing
- [x] I have manually tested the changes

**Deployment**

- [ ] I have added developer documentation (if applicable)
- [ ] Documentation should be added to Zowe Docs
  - If you're an outside contributor, please post in the [#zowe-doc Slack channel](https://openmainframeproject.slack.com/archives/CC961JYMQ) to coordinate documentation.
  - Otherwise, please check with the rest of the squad about any needed documentation before merging.
- [ ] These changes may need ported to the appropriate branches (list here):

### CODE REVIEW

**Review by anaxceron — APPROVED** [review]

**Review by JillieBeanSim — APPROVED** [review]
LGTM! thanks for the fix @traeok

**Review by anaxceron — APPROVED** [review]

**Comment by traeok** [comment]
Need to resolve some failing tests after changing UX behavior, will mark as ready soon

**Review by JillieBeanSim — CHANGES_REQUESTED** [review]
Hey @traeok, I'm not sure what changed but not seeing the pop up like before when clicking file after incorrect creds initiated just the error in file editor.

**Comment by traeok** [comment]
> Hey @traeok, I'm not sure what changed but not seeing the pop up like before when clicking file after incorrect creds initiated just the error in file editor.

Hi Billie, thanks for quickly re-reviewing - can you provide steps to reproduce this scenario? When testing the steps in the PR, I can't reproduce this with data sets, USS or jobs.

**Comment by JillieBeanSim** [comment]
> > Hey @traeok, I'm not sure what changed but not seeing the pop up like before when clicking file after incorrect creds initiated just the error in file editor.
> 
> Hi @JillieBeanSim, thanks for quickly re-reviewing - if possible, can you provide steps to reproduce this scenario? When testing the steps in the PR, I can't reproduce this with data sets, USS or jobs.

Hey @traeok 

1. did filter search MVS with successful creds and expand a data set with members
2. right click profile, prof management, update creds to ones expected to fail
3. click on data set member in expanded tree
4. see error in text editor but no pop up since changes pushed

**Comment by JillieBeanSim** [comment]
> > 1. did filter search MVS with successful creds and expand a data set with members
> > 2. right click profile, prof management, update creds to ones expected to fail
> > 3. click on data set member in expanded tree
> > 4. see error in text editor but no pop up since changes pushed
> > 
> > this scenario worked when I previously approved
> 
> Hmm... I'll have to do some more testing locally. If I follow these same steps, the prompt appears on my machine. Maybe its an OS-specific issue? Other reviewers might be able to confirm this behavior, but the source of the issue is unclear since I can't reproduce with the same steps 😓
> 
> The error will still appear in the text editor until the user clicks "Try again", but the pop up for updating credentials should appear as soon as the error appears in the editor.

yes it worked that way when I tested and approved previously but not getting the pop up after pulling the latest.

**Review by JillieBeanSim — APPROVED** [review]
thanks for addressing my comments, all works great now! Thanks @traeok

File: packages/zowe-explorer/src/trees/ZoweTreeProvider.ts
Comment (zFernand0) [inline packages/zowe-explorer/src/trees/ZoweTreeProvider.ts:1]: Thanks for refactoring and reducing some duplication 🙏

**Review by zFernand0 — APPROVED** [review]
LGTM! 😋 

Thanks for the detailed testing instructions 🙏 
This PR makes recovery much more straightforward 🙏 

I guess I kind of expected a "credentials successfully updated" message after entering my valid creds, but clicking "Try again" worked just fine 😋

**Pull request merged**